### PR TITLE
Rewrite parameters to new classes that allow for more complex validation

### PR DESCRIPTION
**Pull request by Rodentman87**
This implementation of parameters takes a more static approach to parameter types and how they're validated. By creating specific classes for various parameter types, we can validate the parameters before they even reach the handler. This also allows us to have specific validation implementations for the different parameter types. It's also trivial to expand on this as we just create a new class for the new parameter type. The other benefit of this new system is it allows parameters other than strings to be used instead of passing numbers or booleans as strings which get parsed later. 

In this PR, I've also moved around a few things to keep info about methods all in the same place (inside the classes) rather than having some of the info being passed during registration.

Things to follow up on:
* Adding more parameter types (integer, boolean, integer options, lists)
* Fixing some of the validation, such as making sure the float parameter will also accept integers and cast them to floats.

### CODE REVIEW

File: KerbalWebProgram/ApiEndpointsBuiltIn.cs [hunk 1]
```diff
@@ -118,30 +47,29 @@ public class getCelestialBodyData : KWPapi
...
+            };
+            this.Type = "response";
+            this.Name = "Get Celestial Body Data";
+            this.Description = "Allows you to get the Celestial Body Data of a specified Celestial Body";
+            this.Author = "KWP dev team";
+            this.Tags = new List<string> { "CelestialBody" };
         }
 
-        public override ApiResponseData Run(ApiRequestData apiRequestData)
+        public override Dictionary<string, object> Run(dynamic parameters)
```
Comment (NathanPeake) [inline KerbalWebProgram/ApiEndpointsBuiltIn.cs:62]: ApiResponseData should be set here as its for the api to handle that. 

ApiRequestData for the same reason. put the parameters inside of ApiRequestData

File: KerbalWebProgram/KerbalWebProgramMod.cs [hunk 2]
```diff
@@ -257,36 +260,30 @@ static ApiResponseData ApiHandler(ApiRequestData apiRequestData)
             {
                 KWPapi ApiEndpoint = webAPI[apiRequestData.Action];
                 ApiResponseData apiResponseData = new ApiResponseData();
-                apiResponseData.Data = new Dictionary<string, object>();
-                bool hasError = false;
+                apiResponseData.ID = apiRequestData.ID;
```
Comment (NathanPeake) [inline KerbalWebProgram/KerbalWebProgramMod.cs:263]: This should be handled inside of each api.

**Review by NathanPeake — CHANGES_REQUESTED** [review]
ApiRequestData and ApiResponseData is important to have inside of each api the parameters should be put inside of the ApiRequestData

File: KerbalWebProgram/ApiEndpointsBuiltIn.cs [hunk 3]
```diff
@@ -152,27 +80,25 @@ public class getALLCelestialBodyData : KWPapi
...
+            this.Author = "KWP dev team";                
+            this.Tags = new List<string> { "CelestialBody" };
         }
-        public override ApiResponseData Run(ApiRequestData apiRequestData)
+
+        public override Dictionary<string, object> Run(dynamic parameters)
         {
-            ApiResponseData apiResponseData = new ApiResponseData();
-            apiResponseData.ID = apiRequestData.ID;
-            apiResponseData.Type = "response";
```
Comment (NathanPeake) [inline KerbalWebProgram/ApiEndpointsBuiltIn.cs:168]: apiResponseData.Type is used here as each api can have different response types.

**Review by NathanPeake — CHANGES_REQUESTED** [review]
apiResponseData.Type is there for the api to set the type of response it has. its upto the api creator to set this as each api can have its own response set. example being one could have multiple response types for each request, one request could have "response" while another with different input data could have "update".

File: KerbalWebProgram/KerbalWebProgramMod.cs (hunk 2 above)
Comment (NathanPeake) [reply KerbalWebProgram/KerbalWebProgramMod.cs:263]: Line 263 (github did not specify the line I was talking about)

**Comment by Rodentman87** [comment]
I'm not sure semantically what the difference here between "response" and "update" is, but the method handlers really shouldn't be directly touching the response object. The problem with having them all make the response themselves is that it makes things very prone to breakage if we decide to modify the way responses look later down the line. If we decide to add a new property, then *every single api handler* (including ones made by other mods) has to be updated to set that property. We should be trying to handle as much as we can outside of the actual method handlers so that the handlers can remain as simple as possible.

I'm also not really sure what the point of passing the entire request data into each handler is for. The only thing actually relevant to the handler is the parameters.

**Comment by NathanPeake** [comment]
The point of giving the ApiResponseData and ApiRequestData is to allow the api to get and set what it needs to.
like the response type. not for us to do beforehand in the main server.

The response type "update" was just an example. the api should set its own response as each api is different. as another example an api could be streaming different types of data so it might set the response type to "cockpit" each api also has its own documentation to show how to handle the response type. (example from #24)

This is to replicate the equivalent of (as an example) the nodejs versions of http.ServerResponse and http.ClientRequest. where the API's can change the type of response and access the request that has happened. 

Regarding the properties issue, we should limit updating these to major version changes since that is a major change.

**Comment by Rodentman87** [comment]
Then I guess I'm misunderstanding what the response type is. Because I was under the impression that the response types were going to be "response" for a direct response to a request, "error" for an error response to a request, and "event" for server-sent events once the websocket stuff is ready. What is a "cockpit" response type and why is that important to someone consuming the api? Realistically even the error response type could just be a "response" with errors instead of data.

As for trying to mimic the nodejs type api where it provides raw request and response data, why? Like what is the point of not handling the rest of the stuff automatically. What is the advantage of making every single handler set the response's ID?

As for the breaking changes thing, we don't even have any sort of 3rd party API or anything, nor have we had the documentation for more than a few hours. While it would be a breaking change, why are you trying to avoid that when it provides a lot of benefits? If this change requires a new major version, then we should do so rather than continuing to build on the current foundation.

**Comment by NathanPeake** [comment]
ApiResponseData and ApiRequestData is for the apis to use if it was for the server I would have called them ServerResponseData and ServerRequestData

"cockpit" response type was just an example of what an api could set it as. the documentation generated would cover what it is and what it does. its upto the api to define it. 

the node js thing was an example.

We do have a third party api that was built into it right at the start, hence why ApiResponseData and ApiRequestData are a thing for the apis to use. the inbuilt api uses the third party api system. 

I want to merge this into it because everything else is good except the parts I have highlighted.

I think your getting very confused about what Im saying and about what this project is. everything you have done is good,  ApiResponseData and ApiRequestData just need adding back to where they was.

**Comment by Rodentman87** [comment]
When I had originally suggested this structure for the API, the type was supposed to say which response type was being sent, not the type structure of the actual data. The reason for `error`, `response`, and `event` was for telling clients which type of data they were being sent (either a response to a request they made, an error response to a request they made, or a server-sent event) so they could sort it appropriately. This is much less useful in the HTTP API, but is incredibly important when trying to parse messages coming over a websocket. I don't think using `type` for the returned data type makes much sense as that will almost always be constant for each method in the API. Having the structure of the returned data documented is important, but I don't think that's the place to put it. That type should be in the documentation, not the returned data.

I understand that using the ApiRequestData and ApiResponseData is more similar to writing normal server handlers, but we've already started handling the actual request/response. By leaving things like copying over the ID up to the handler, when all the handler should really need to care about is the action its taking or the data its returning, it makes it really easy to cause issues or write things in an inconsistent manner. This is a pretty major footgun here for anyone not familiar with the API trying to make their own handlers. I would understand having passing the full request and returning the full response if the API was more freeform, but the actual API is much more rigid than that.

**Comment by NathanPeake** [comment]
The api is not rigid. and the changes you want to make with ApiResponseData and ApiRequestData would make it rigid.

**Comment by Rodentman87** [comment]
I guess I misunderstood what the intended API structure was then, I'll get that changed

**Comment by NathanPeake** [comment]
Thanks. I will approve when done.

**Comment by Rodentman87** [comment]
If the `type` is supposed to be the data type of the responses, then all of the current methods should be updated to have a proper response type as well. I'm not sure what you're wanting those to be, so I'll leave them as they are for now

**Review by NathanPeake — APPROVED** [review]

**Comment by Rodentman87** [comment]
Ok I think resolved the merge conflicts correctly, and I don't think I missed any methods when updating to the new format, but if someone could double check my work just to be sure that would be great.

**Review by NathanPeake — APPROVED** [review]

**Pull request merged**